internal/backend/telemetry/telemetry.go

```diff
@@ -0,0 +1,143 @@
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.20.0"
+	"go.uber.org/zap"
+)
+
+type Config struct {
+	Enabled     bool   `koanf:"enabled"`
+	ServiceName string `koanf:"service_name"`
+	Endpoint    string `koanf:"endpoint"`
+}
+
+var Configs = configset.Set[Config]{
+	Default: &Config{Enabled: true, ServiceName: "ak", Endpoint: "localhost:4318"},
+	Dev:     &Config{Enabled: false, ServiceName: "ak", Endpoint: "localhost:4318"},
+}
+
+func toMetricAttrs(attrs map[string]string) []attribute.KeyValue {
+	var result []attribute.KeyValue
+	for k, v := range attrs {
+		result = append(result, attribute.String(k, v))
+	}
+	return result
+}
+
+type Metric interface {
+	Update(value int64, attrs map[string]string)
+}
+
+type Counter struct {
+	counter api.Int64Counter
+}
+
+type UpDownCounter struct {
+	counter api.Int64UpDownCounter
+}
+
+func (m Counter) Update(value int64, attrs map[string]string) {
+	// REVIEW: should we check/report if we get a negative value
+	if enabled {
+		m.counter.Add(context.Background(), value, api.WithAttributes(toMetricAttrs(attrs)...))
+	}
+}
+
+func (m UpDownCounter) Update(value int64, attrs map[string]string) {
+	if enabled {
+		m.counter.Add(context.Background(), value, api.WithAttributes(toMetricAttrs(attrs)...))
+	}
+}
+
+var enabled = false
+
+func metricName(field reflect.Value, fieldType reflect.StructField) string {
+	metricName := fieldType.Name
+	metricName = strings.ReplaceAll(metricName, "_", ".")
+	metricName = "ak." + strings.ToLower(metricName)
+	switch field.Interface().(type) {
+	case Counter:
+		metricName = metricName + ".counter"
+	case UpDownCounter:
+		metricName = metricName + ".gauge"
+	}
+	return metricName
+}
+
+func Init(z *zap.Logger, cfg *Config) {
+	enabled = cfg.Enabled
+	if !enabled {
+		z.Warn("metrics are disabled")
+	}
+
+	// REVIEW: insecure? port? maybe over GRPC?
+	// REVIEW: should we use AK env or maybe standard OTEL onses?
+	//         - OTEL_EXPORTER_OTLP_ENDPOINT or OTEL_EXPORTER_OTLP_METRICS_ENDPOINT
+	exporter, err := otlpmetrichttp.New(
+		context.Background(),
+		otlpmetrichttp.WithInsecure(),
+		otlpmetrichttp.WithEndpoint(cfg.Endpoint),
+		// otlpmetrichttp.WithURLPath("/"), // /v1/metrics by default, unless set
+	)
+	if err != nil {
+		z.Error("failed to create metric exporter: %v", zap.Error(err))
+	}
+
+	schemaURL := "https://opentelemetry.io/schemas/1.1.0"
+	resourceAttrs := resource.NewWithAttributes(
+		schemaURL,
+		semconv.ServiceNameKey.String(cfg.ServiceName),
+	)
+	fmt.Println("with service name", cfg.ServiceName)
+
+	// REVIEW: consider using controller?
+	meterProvider := metric.NewMeterProvider(
+		metric.WithReader(metric.NewPeriodicReader(exporter)),
+		metric.WithResource(resourceAttrs),
+	)
+
+	otel.SetMeterProvider(meterProvider) // set global meter provider
+	meter := otel.GetMeterProvider().Meter("ak")
+
+	metricsValue := reflect.ValueOf(&Metrics).Elem()
+	metricsType := metricsValue.Type()
```

Comment:
I do not understand the purpose of this code.

Reply:
the purpose of this code is to initialize all metrics defined by the user.

Reply:
i am very wary of adding this code with reflection here. i have no idea how it works. i have no idea why you even use reflection here. either add very clear comments about why it does that and how it does that, or find a better way.

Reply:
what i mean is that you need to justify the use of reflection here. and I'm not sure it's possible.